state_version.go

```diff
@@ -243,6 +273,46 @@ func (s *stateVersions) Create(ctx context.Context, workspaceID string, options
 	return sv, nil
 }
 
+func (s *stateVersions) putURL(ctx context.Context, putURL string, data []byte) error {
+	reader := bytes.NewReader(data)
+	req, err := s.client.NewRequest("PUT", putURL, reader)
+	if err != nil {
+		return err
+	}
+
+	return req.Do(ctx, nil)
+}
+
+func (s *stateVersions) Upload(ctx context.Context, workspaceID string, options StateVersionUploadOptions) (*StateVersion, error) {
+	if err := options.valid(); err != nil {
+		return nil, err
+	}
+
+	sv, err := s.Create(ctx, workspaceID, options.StateVersionCreateOptions)
+	if err != nil {
+		if strings.Contains(err.Error(), "param is missing or the value is empty: state") {
+			return nil, ErrStateVersionUploadNotSupported
+		}
+	}
+
+	g, _ := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		return s.putURL(ctx, sv.UploadURL, options.RawState)
+	})
+	if options.RawJSONState != nil {
+		g.Go(func() error {
+			return s.putURL(ctx, sv.JSONUploadURL, options.RawJSONState)
+		})
+	}
```

Comment:
Should ensure `sv.JSONUploadURL` and `sv.UploadURL` are not empty strings?

Reply:
I don't think it's possible since the state version was just created with no content.

Reply:
This turned into an adventure! See my notes in the PR description under "Refactor Notes"